File: ctci/tmp/chapter_3_stacks_and_queues/python/3.1-three_in_one.py
class FixedMultiStack:
    def __init__(self, stack_size: int):
        """
        Fixed multi-stack constructor.
        """
        self.number_of_stacks = 3
        self.stack_capacity = stack_size
        self.values = [0] * stack_size * 3
        # Array of sizes of sub-arrays.
        self.sizes = [0] * 3

    def push(self, stack_num: int, value: int):
        """
        Pushes an element into a given stack.
        """
        if self.is_full(stack_num):
            raise Exception("stack is full")
        self.sizes[stack_num] += 1
        index = self.index_of_top(stack_num)
        self.values[index] = value

    def pop(self, stack_num: int) -> int:
        """
        Pops the top element off a fixed multi-stack.
        """
        if self.is_empty(stack_num):
            raise Exception("stack is empty")
        index = self.index_of_top(stack_num)
        value = self.values[index]
        self.values[index] = 0
        self.sizes[stack_num] -= 1
        return value

    def is_full(self, stack_num: int) -> bool:
        """
        Determines if a fixed multi-stack is full or not.
        """
        return self.sizes[stack_num] == self.stack_capacity

    def is_empty(self, stack_num: int) -> bool:
        """
        Determines if a fixed multi-stack is empty or not.
        """
        return self.sizes[stack_num] == 0

    def index_of_top(self, stack_num: int) -> int:
        """
        Returns the top index of a given stack.
        """
        offset = stack_num * self.stack_capacity
        index = self.sizes[stack_num]
        return offset + index - 1


if __name__ == "main":
    fms = FixedMultiStack(3)
    print(fms.values)  # [0, 0, 0, 0, 0, 0, 0, 0, 0]
    print(fms.index_of_top(0))  # -1
    print(fms.index_of_top(1))  # 2
    print(fms.index_of_top(2))  # 5
    fms.push(0, 5)
    fms.push(0, 10)
    fms.push(0, 15)
    # fms.push(0, 20) exception
    fms.push(1, 6)
    print(fms.values)  # [5, 10, 15, 6, 0, 0, 0, 0, 0]
    print(fms.is_full(0))  # True
    print(fms.is_empty(1))  # False
    print(fms.is_empty(2))  # True
    print(fms.pop(0))  # 15
    print(fms.values)  # [5, 10, 15, 6, 0, 0, 0, 0, 0]
    print(fms.sizes)  # [2, 1, 0]
    fms.push(0, 25)
    print(fms.values)  # [5, 10, 25, 6, 0, 0, 0, 0, 0]
    # fms.pop(2) exception

File: ctci/tmp/chapter_3_stacks_and_queues/cpp/3.1-three_in_one.cpp
#include <iostream>

using namespace std;

class FixedMultiStack
{
public:
  int numOfStacks = 3;
  int stackCapacity;
  int *values;
  int size;
  /* Array of sizes of sub-arrays. */
  int sizes[3] = {0};

  FixedMultiStack(int stackSize);
  void printStack(int *arr, int len);
  void push(int stackNum, int value);
  int pop(int stackNum);
  int indexOfTop(int stackNum);
  bool isEmpty(int stackNum);
  bool isFull(int stackNum);
};

/**
 * Fixed multi-stack constructor.
 */
FixedMultiStack::FixedMultiStack(int stackSize)
{
  this->stackCapacity = stackSize;
  this->size = stackSize * this->numOfStacks;
  this->values = new int[size]{0};
};

/**
 * Prints a given stack in a fixed multi-stack.
 */
void FixedMultiStack::printStack(int *arr, int len)
{
  for (int i = 0; i < len - 1; i++)
    cout << arr[i] << ", ";
  cout << arr[i] << endl;
}

/**
 * Pushes an element into a given stack.
 */
void FixedMultiStack::push(int stackNum, int value)
{
  if (stackNum < 0 || stackNum > 2)
    throw runtime_error("invalid index");
  if (this->isFull(stackNum))
    throw runtime_error("stack is full");
  this->sizes[stackNum]++;
  int index = this->indexOfTop(stackNum);
  this->values[index] = value;
}

/**
 * Pops the top element off a fixed multi-stack.
 */
int FixedMultiStack::pop(int stackNum)
{
  if (stackNum < 0 || stackNum > 2)
    throw runtime_error("invalid index");
  if (this->isEmpty(stackNum))
    throw runtime_error("stack is empty");
  int index = this->indexOfTop(stackNum);
  int value = this->values[index];
  this->values[index] = 0;
  this->sizes[stackNum]--;
  return value;
}

/**
 * Returns the top index of a given stack.
 */
int FixedMultiStack::indexOfTop(int stackNum)
{
  int offset = stackNum * this->stackCapacity;
  int index = this->sizes[stackNum];
  return offset + index - 1;
}

/**
 * Determines if a fixed multi-stack is empty or not.
 */
bool FixedMultiStack::isEmpty(int stackNum)
{
  return this->sizes[stackNum] == 0;
}

/**
 * Determines if a fixed multi-stack is full or not.
 */
bool FixedMultiStack::isFull(int stackNum)
{
  return this->sizes[stackNum] == this->stackCapacity;
}

int main()
{
  FixedMultiStack *fms;

  fms = new FixedMultiStack(4);
  fms->printStack(fms->values, fms->size);
  /* 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 */
  fms->printStack(fms->sizes, fms->numOfStacks);
  /* 0, 0, 0 */
  cout << fms->indexOfTop(0) << endl; // -1
  cout << fms->indexOfTop(1) << endl; // 3
  cout << fms->indexOfTop(2) << endl; // 7
  cout << fms->isEmpty(2) << endl;    // 1
  fms->push(1, 32);
  fms->push(1, 64);
  fms->push(1, 128);
  fms->push(1, 256);
  // fms->push(1, 256); exception
  cout << fms->isEmpty(1) << endl; // 0
  cout << fms->isFull(1) << endl;  // 1
  fms->printStack(fms->values, fms->size);
  /* 0, 0, 0, 0, 32, 64, 128, 256, 0, 0, 0, 0 */
  fms->printStack(fms->sizes, fms->numOfStacks);
  /* 0, 4, 0 */
  cout << fms->pop(1) << endl; // 256
  // fms->pop(2); exception
  fms->printStack(fms->values, fms->size);
  /* 0, 0, 0, 0, 32, 64, 128, 0, 0, 0, 0, 0 */
  fms->printStack(fms->sizes, fms->numOfStacks);
  /* 0, 3, 0 */
  delete fms->values;
  delete fms;

  return 0;
}